### DISCUSSION TITLE
Autocompletion based on JSON schema

### DISCUSSION THREAD

**Discussion by rxxy** [josdejong/svelte-jsoneditor#366] [Ideas]

**Comment by josdejong** [comment]
what?

**Comment by rxxy** [comment]
![image](https://user-images.githubusercontent.com/14834156/184650927-80c29eec-e942-4b91-b9d9-0bd5847c871c.png)

Like intellij idea, when schema is specified, input 'n' can be completed to 'name'

**Comment by josdejong** [comment]
Ah, now I understand you, you mean auto-completion based on a JSON Schema. That's not yet supported but would be very nice to have.

Anyone interested in picking this up?

**Comment by developergunny** [comment]
I need!

**Comment by AlexRMU** [comment]
Nice to have

**Comment by Kasheftin** [comment]
Please share more details about autocomplete in svelte-jsoneditor / vanilla-jsoneditor.
As I understand, the previous version, jsoneditor, supports some kind of autocomplete, https://codesandbox.io/s/z65fp?file=/src/App.vue. The newer version uses @codemirror/autocomplete somewhere, but there're no examples and there's nothing about autocomplete in the docs.

Our use case is that we have a schema for createAjvValidator with enums, validation works just fine, but we want to have autocomplete functionality both for keys and enum values.

**Comment by josdejong** [comment]
@Kasheftin There is no autocompletion for JSON Schema implemented yet in `svelte-jsoneditor` like there is in `jsoneditor`, hence this issue. Help would be welcome.

**Comment by jy95** [comment]
@josdejong If it could help, I saw this project https://github.com/acao/codemirror-json-schema . They might not have the ctrl-space trick users known from monaco-editor but it is better than nothing from what I saw on https://codemirror-json-schema.netlify.app/.

![image](https://github.com/josdejong/svelte-jsoneditor/assets/9306961/27367de6-4d63-4dad-9367-c7abcf62bf58)

They seem to have documented something for [custom usage](https://github.com/acao/codemirror-json-schema#custom-usage) that could be added to your [.create call](https://github.com/josdejong/svelte-jsoneditor/blob/main/src/lib/components/modes/textmode/TextMode.svelte#L543)

```ts
// This approach allows you to configure the json mode and parse linter, as well as our linter, hovers, etc more specifically.

import { EditorState } from "@codemirror/state";
import { linter } from "@codemirror/lint";
import { hoverTooltip } from "@codemirror/view";
import { json, jsonLanguage, jsonParseLinter } from "@codemirror/lang-json";
import { jsonCompletion jsonSchemaLinter, jsonSchemaHover } from "codemirror-json-schema";

const schema = {
  type: "object",
  properties: {
    example: {
      type: "boolean",
    },
  },
};

const state = EditorState.create({
  doc: `{ "example": true }`,
  extensions: [
    json(),
    linter(jsonParseLinter(), {
      // default is 750ms
      delay: 300
    }),
    linter(jsonSchemaLinter(schema)),
    jsonLanguage.data.of({
      autocomplete: jsonCompletion(schema),
    }),
    hoverTooltip(jsonSchemaHover(schema)),
  ];
})
```

Notice : It has currently an open MR called ".js file imports and schema loading as statefield" so that you don't have to provide the schema in the create step

**Comment by josdejong** [comment]
Thanks for sharing @jy95 , this can be very useful indeed!

Anyone interested in working this out for `svelte-jsoneditor`?

**Comment by jy95** [comment]
@josdejong I set up a [stackblitz](https://stackblitz.com/edit/github-vhecvd?file=src%2Flib%2Fcomponents%2Fmodes%2Ftextmode%2FTextMode.svelte) for POC that is uncomplete for the time being , mostly because I am not used to the prop drilling in Svelte codebase (`export let schema: any`).

**Comment by josdejong** [comment]
Thanks @jy95!

**Comment by jy95** [comment]
@josdejong In the meantime, [codemirror-json-schema](https://github.com/acao/codemirror-json-schema) got one interesting [release](https://github.com/acao/codemirror-json-schema/releases/tag/v0.5.0) . What do you think ?

```ts
extensions: [
  json(),
  linter(jsonParseLinter()),
  linter(jsonSchemaLinter(), {
    needsRefresh: handleRefresh,
  }),
  jsonLanguage.data.of({
    autocomplete: jsonCompletion(),
  }),
  hoverTooltip(jsonSchemaHover()),
  // this is where we pass the schema!
  // very important!!!!
  stateExtensions(schema),
];
```

**Comment by josdejong** [comment]
That looks very promising indeed! Either to just use `codemirror-json-schema`, or to see how they solved autocompletion.

**Comment by josdejong** [comment]
Moving this idea to the [Discussions > Ideas](https://github.com/josdejong/svelte-jsoneditor/discussions/categories/ideas) section since we are not actively working on it. Help implementing this would be very welcome.

**Comment by stereokai** [comment]
@josdejong Has this already been implemented/figured out? Thanks 🙏
